pallets/router/src/tests.rs

```diff
@@ -0,0 +1,119 @@
+
+#[test]
+fn too_many_or_too_less_routes_should_not_work() {
+    ExtBuilder::default().build().execute_with(|| {
+        let routes_11 = core::iter::repeat((0, DOT, XDOT)).take(11).collect();
+        assert_noop!(
+            AMMRoute::trade(Origin::signed(ALICE), routes_11, 1, 2, 3),
+            Error::<Runtime>::ExceedMaxLengthRoute
+        );
+
+        assert_noop!(
+            AMMRoute::trade(Origin::signed(ALICE), vec![], 1, 2, 3),
+            Error::<Runtime>::EmptyRoute
+        );
+    });
+}
+
+#[test]
+fn duplicated_routes_should_not_work() {
+    ExtBuilder::default().build().execute_with(|| {
+        let dup_routes = vec![(0, DOT, XDOT), (0, DOT, XDOT)];
+        assert_noop!(
+            AMMRoute::trade(Origin::signed(ALICE), dup_routes, 1, 2, 3),
+            Error::<Runtime>::DuplicatedRoute
+        );
+    });
+}
+
+#[test]
+fn too_low_balance_should_not_work() {
+    ExtBuilder::default().build().execute_with(|| {
+        let dup_routes = vec![(0, DOT, XDOT)];
+        assert_noop!(
+            AMMRoute::trade(Origin::signed(ALICE), dup_routes, 0, 0, 3),
+            Error::<Runtime>::ZeroBalance
+        );
+    });
+}
+
+#[test]
+fn too_small_expiry_should_not_work() {
+    ExtBuilder::default().build().execute_with(|| {
+        let dup_routes = vec![(0, DOT, XDOT), (1, DOT, XDOT)];
+        let current_block_num = 4;
+        run_to_block(current_block_num);
+
+        assert_noop!(
+            AMMRoute::trade(
+                Origin::signed(ALICE),
+                dup_routes,
+                1,
+                2,
+                current_block_num - 1
+            ),
+            Error::<Runtime>::TooSmallExpiry
+        );
+    });
+}
+
+#[test]
+fn trade_should_work() {
+    ExtBuilder::default().build().execute_with(|| {
+        // create pool and add liquidity
+        assert_ok!(DOT2XDOT::add_liquidity(
+            Origin::signed(DAVE),
+            (DOT, XDOT),
+            (100_000_000, 100_000_000),
+            (99_999, 99_999),
+        ));
+
+        // check that pool was funded correctly
+        assert_eq!(DOT2XDOT::pools(XDOT, DOT).unwrap().base_amount, 100_000_000); // XDOT
+        assert_eq!(
+            DOT2XDOT::pools(XDOT, DOT).unwrap().quote_amount,
+            100_000_000
+        ); // DOT
+
+        // calculate amount out
+        assert_ok!(AMMRoute::trade(
+            Origin::signed(ALICE),
+            vec![(0, DOT, XDOT)],
+            1_000,
+            980,
+            1
+        ));
+
+        // Check Alice should get 994
+        assert_eq!(Currencies::free_balance(XDOT, &ALICE), 994);
+
+        // pools values should be updated - we should have less XDOT
+        assert_eq!(DOT2XDOT::pools(XDOT, DOT).unwrap().base_amount, 99_999_006);
+
+        // pools values should be updated - we should have more DOT in the pool
+        assert_eq!(
+            DOT2XDOT::pools(XDOT, DOT).unwrap().quote_amount,
+            100_000_998
+        );
+    })
+}
```

Comment:
This test is great however it doesn't cover the case of a cross pool trade (where we have more than one route). Let's add one to cover this.